lib/confex/type.ex

```diff
@@ -0,0 +1,85 @@
+defmodule Confex.Type do
+  @moduledoc """
+  This module provides API to parse strings and cast them to Elixir types.
+  """
+  @type value :: String.t | nil
+  @type type :: :string
+              | :integer
+              | :float
+              | :boolean
+              | :atom
+              | :module
+              | :list
+
+  @boolean_true ["true", "1", "yes"]
+  @boolean_false ["false", "0", "no"]
+  @list_separator ","
+
+  @doc """
+  Parse string and cast it to Elixir type.
+  """
+  @spec cast(value :: value, type :: type) :: {:ok, any()} | {:error, String.t}
+  def cast(nil, _),
+    do: {:ok, nil}
+  def cast(value, :string),
+    do: {:ok, value}
+  def cast(value, :module),
+    do: {:ok, Module.concat([value])}
+  def cast(value, :integer) do
+    case Integer.parse(value) do
+      {integer, ""} ->
+        {:ok, integer}
+
+      {_integer, remainder_of_binary} ->
+        {:error, "can not cast #{inspect value} to Integer, " <>
+                 "result contains binary remainder #{remainder_of_binary}"}
+      :error ->
+        {:error, "can not cast #{inspect value} to Integer"}
+    end
+  end
+  def cast(value, :float) do
+    case Float.parse(value) do
+      {float, ""} ->
+        {:ok, float}
+
+      {_float, remainder_of_binary} ->
+        {:error, "can not cast #{inspect value} to Float, " <>
+                 "result contains binary remainder #{remainder_of_binary}"}
+
+      :error ->
+        {:error, "can not cast #{inspect value} to Float"}
+    end
+  end
+  def cast(value, :atom) do
+    result =
+      value
+      |> String.to_char_list()
+      |> List.to_atom()
+
+    {:ok, result}
+  end
+
```

Comment:
By most of the style guides, in different languages - single whitespace in the end of file is preferred.